File: 90. Subsets II/90. Subsets II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        map<int, int> m;
        for (int n : nums) {
            map<int, int>::iterator it = m.find(n);
            if (it != m.end()) {
                it->second++;
            } else {
                m.insert(pair<int, int>(n, 1));
            }
        }
        
        
        vector<vector<int>> res(1, vector<int>(0));
        
        for (map<int, int>::iterator it = m.begin(); it != m.end(); it++) {
            vector<vector<int>> temp;
            for (vector<int>& s : res) {
                for (int i = 0; i <= it->second; i++) {
                    vector<int> t(s);
                    for (int j = 0; j < i; j++) {
                        t.push_back(it->first);
                    }
                    temp.push_back(t);
                }
            }
            res = temp;
        }
        
        return res;
        
    }
};

File: 90. Subsets II/90. Subsets II.rs
use std::collections::HashMap;

impl Solution {
    pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut m: HashMap<i32, i32> = HashMap::new();
        for n in nums {
            if let Some(count) = m.get_mut(&n) {
                *count += 1;
            } else {
                m.insert(n, 1);
            }
        }
        let mut res: Vec<Vec<i32>> = vec![vec![]];
        let mut next: Vec<Vec<i32>> = Vec::new();
        
        for (n, count) in m {
            for ss in res {
                for i in 0..=count {
                    let mut ss = ss.clone();
                    for j in 0..i {
                        ss.push(n);                        
                    }
                    next.push(ss);
                }
            }
            res = next;
            next = Vec::new();
        }
        return res;
    }
}